hi-ml/src/health_ml/utils/checkpoint_handler.py

```diff
@@ -69,10 +76,56 @@ def get_checkpoint_to_test(self) -> Path:
+            if downloaded_checkpoint is not None:
+                logging.info(f"Using a checkpoint found in the AzureML run: {downloaded_checkpoint}")
+                return downloaded_checkpoint
+            raise FileNotFoundError("No inference checkpoint file found locally nor in AzureML.")
         elif self.trained_weights_path:
             # Model was not trained, check if there is a local weight path.
             logging.info(f"Using pre-trained weights from {self.trained_weights_path}")
             return self.trained_weights_path
         raise ValueError("Unable to determine which checkpoint should be used for testing.")
+
+    def get_relative_inference_checkpoint_path(self) -> Path:
+        """Returns the path of the model's inference checkpoint, relative to the container's output folder."""
+        expected_checkpoint_path = self.container.get_checkpoint_to_test()
+        try:
+            # Find the Unix-style path of the checkpoint relative to the outputs folder, this will be the path
+            # where we find the file in AzureML.
+            return expected_checkpoint_path.relative_to(self.container.outputs_folder)
+        except ValueError:
+            raise ValueError("Inference checkpoint path should be relative to the container's output folder. "
+                             f"Checkpoint path: {expected_checkpoint_path}, "
+                             f"output folder: {self.container.outputs_folder}")
+
+    def download_inference_checkpoint(self, download_folder: Optional[Path] = None) -> Optional[Path]:
+        """
+        For low-priority preemption that occured after training, try to download the inference checkpoint if that
+        is available in the AzureML run from a previous incarnation of the job. Downloads go into the `download_folder`.
+
+        :param download_folder: The folder where the checkpoints should be downloaded to. If not provided, use a
+            temp folder.
+        :return: The path to a downloaded inference checkpoint.
+        """
+        # This logic will only trigger in AzureML. Download should only happen once per node.
+        if is_running_in_azure_ml() and is_global_rank_zero():
+            download_folder = download_folder or Path(tempfile.mkdtemp())
+            inference_checkpoint_azureml_path = (
+                Path(DEFAULT_AML_UPLOAD_DIR) / self.get_relative_inference_checkpoint_path()
+            ).as_posix()
+            highest_epoch_checkpoint = download_highest_epoch_checkpoint(
+                run=RUN_CONTEXT,
+                checkpoint_suffix=inference_checkpoint_azureml_path,
+                output_folder=download_folder)
+            if highest_epoch_checkpoint is None:
+                logging.info("No inference checkpoint was found in the AzureML run.")
+                return None
+            logging.info("An inference checkpoint was found in the AzureML run.")
+            return highest_epoch_checkpoint
+        return None
```

Comment:
Unnecessary - can be removed

Reply:
you mean the logging line? I'd rather be generous here with logging. If something goes wrong with the logic here, we need to be able to easily trace back what happened because we can't reproduce preemptions, and may have to code up fixes purely on the basis of these diagnostics.

Reply:
Sorry should have been clearer, my comment is about line 131: the `return None` is superfluous as the function will return `None` if it reaches that line anyway

Reply:
I see... yes, strictly speaking the `return None` is superfluous, but that's now the functional-programming-Anton in me saying that you must return something in all codepaths for clarity.